src/hardware/input/keyboard.cpp

```diff
@@ -8,492 +11,700 @@
  */
 
-#include "dosbox.h"
 #include "keyboard.h"
+#include "dosbox.h"
+
+#include <array>
 
-#include "bios.h"
 #include "bitops.h"
-#include "inout.h"
-#include "mem.h"
-#include "mixer.h"
+#include "checks.h"
+#include "cpu.h"
+#include "intel8042.h"
+#include "intel8255.h"
 #include "pic.h"
 #include "support.h"
 #include "timer.h"
 
-#define KEYBUFSIZE 32
-#define KEYDELAY   0.300 // Considering 20-30 khz serial clock and 11 bits/char
-
 using namespace bit::literals;
 
-enum KeyCommands {
-	CMD_NONE,
-	CMD_SETLEDS,
-	CMD_SETTYPERATE,
-	CMD_SETOUTPORT
+CHECK_NARROWING();
+
+// Emulates the PS/2 keybaord, as seen by the Intel 8042 microcontroller.
+
+// Reference:
+// - https://wiki.osdev.org/PS/2_Keyboard
+// - https://stanislavs.org/helppc/keyboard_commands.html
+// - https://kbd-project.org/docs/scancodes/scancodes.html
+// - https://homepages.cwi.nl/~aeb/linux/kbd/scancodes.html
+// - http://www-ug.eecg.toronto.edu/msl/nios_devices/datasheets/PS2%20Keyboard%20Protocol.htm
+
+#ifdef ENABLE_SCANCODE_SET_3
+#	undef HAS_SCANCODE_SET_2_OR_3
+#	define HAS_SCANCODE_SET_2_OR_3
+#endif // ENABLE_SCANCODE_SET_3
+#ifdef ENABLE_SCANCODE_SET_2
+#	undef HAS_SCANCODE_SET_2_OR_3
+#	define HAS_SCANCODE_SET_2_OR_3
+#endif // ENABLE_SCANCODE_SET_2
+
+static constexpr uint8_t buffer_size = 8; // in scancodes
+
+enum class Command : uint8_t { // PS/2 keyboard port commands
+	None                 = 0x00,
+	SetLEDs              = 0xed,
+	Echo                 = 0xee,
+	CodeSet              = 0xf0,
+	Identify             = 0xf2,
+	SetTypeRate          = 0xf3,
+	ClearEnable          = 0xf4,
+	DefaultDisable       = 0xf5,
+	ResetEnable          = 0xf6,
+	Set3AllTypematic     = 0xf7,
+	Set3AllMakeBreak     = 0xf8,
+	Set3AllMakeOnly      = 0xf9,
+	Set3AllTypeMakeBreak = 0xfa,
+	Set3KeyTypematic     = 0xfb,
+	Set3KeyMakeBreak     = 0xfc,
+	Set3KeyMakeOnly      = 0xfd,
+	Resend               = 0xfe,
+	Reset                = 0xff,
 };
 
+// Internal keyboard scancode buffer
+
+std::vector<uint8_t> buffer[buffer_size] = {};
+
+static bool buffer_overflowed  = false;
+static size_t buffer_start_idx = 0;
+static size_t buffer_num_used  = 0;
+
+// Key repetition mechanism data
+
 static struct {
-	uint8_t buffer[KEYBUFSIZE];
-	Bitu used;
-	Bitu pos;
-	struct {
-		KBD_KEYS key;
-		Bitu wait;
-		Bitu pause,rate;
-	} repeat;
-	KeyCommands command;
-	uint8_t p60data;
-	bool p60changed;
-	bool active;
-	bool scanning;
-	bool scheduled;
-} keyb;
-
-static void KEYBOARD_SetPort60(uint8_t val) {
-	keyb.p60changed=true;
-	keyb.p60data=val;
-	if (machine==MCH_PCJR) PIC_ActivateIRQ(6);
-	else PIC_ActivateIRQ(1);
+	KBD_KEYS key   = KBD_NONE; // key which went typematic
+	uint16_t wait  = 0;
+	uint16_t pause = 0;
+	uint16_t rate  = 0;
+
+} repeat;
+
+struct Set3CodeInfoEntry {
+	bool is_enabled_typematic = true;
+	bool is_enabled_make      = true;
+	bool is_enabled_break     = true;
+};
+
+static std::array<Set3CodeInfoEntry, UINT8_MAX + 1> set3_code_info = {};
```

Comment:
`UINT8_MAX + 1` can be found in a few places in this file. I suggest to introduce a constant called `NumScancodes` or whatever and use that instead.

Reply:
Actually, in just 2 places. `intel8042.cpp` also uses this, but for a different purpose. Nevertheless, `constexpr` added.